### ISSUE TITLE
dtls_udp_echo server : DTLSv1_listen() still return 1 while client has been disconnected

### ISSUE THREAD

**Issue by fantansy** [nplab/DTLS-Examples#8]
do anybody try the daemon?
I built and run dtls_udp_echo , on client side, socket can be set up and recv echo response successfully. 
but on server side, DTLSv1_listen() still return 1 then new alloc thread to handle it. 
my openssl version:
openssl-1.1.0j

**Comment by weinrank** [comment]
Can you provide additional information?
- OS
- arguments
- output

Thanks!

**Comment by liuqun** [comment]
```
# Start echo server
./dtls_udp_echo -L 127.0.0.1 -v
```

```
# Start standard openssl dtls client to talk with the echo server
openssl s_client -dtls1_2 -CAfile certs/ca-cert.pem   -key certs/client-key.pem -cert certs/client-cert.pem 127.0.0.1:23232
```
liuqun@vmware:~/DTLS-Examples/src$ which openssl
/usr/bin/openssl

liuqun@vmware:~/DTLS-Examples/src$ /usr/bin/openssl version
OpenSSL 1.1.1b  26 Feb 2019

liuqun@vmware:~/DTLS-Examples/src$ ldd /usr/bin/openssl 
	linux-vdso.so.1 (0x00007ffc44582000)
	libssl.so.1.1 => /lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f37a50c5000)
	libcrypto.so.1.1 => /lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f37a4df7000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f37a4dd6000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f37a4beb000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f37a4be5000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f37a522f000)

```

**Comment by liuqun** [comment]
On Ubuntu 19.04, the behavior of DTLSv1_listen() changes if 2 clients are connecting concurrently.

* OS: Ubuntu 19.04 with openssl 1.1.1b (via `sudo apt install libssl-dev`)
* dtls_udp_echo arguments:
```
./dtls_udp_echo -L 127.0.0.1 -v
```

* s_client arguments:
```
openssl s_client -dtls1_2 -CAfile ../ca-cert.pem   -key certs/client-key.pem -cert certs/client-cert.pem -connect 127.0.0.1:23232
```

**Comment by liuqun** [comment]
`openssl s_client -dtls`'s default behavior:
 - retry sending "Client Hello" request again after 1, 2, 4, 8, 16, 32 seconds if the DTLS server did not respond anything.

**Comment by weinrank** [comment]
Hey @liuqun,

I've tested the issue on three operating systems: OSX, FreeBSD 12 and Ubuntu 19.04
I just ran your `dtls_udp_echo` server with openssl client example several times and observed the output.

### Operating Systems
- Ubuntu 19.04 with OpenSSL_1_1_1-stable from GIT (commit d63d841fb510a920275c66d3e486089c5c718797)
- OSX 10.14.5 with OpenSSL 1.1.1c  28 May 2019 from Homebrew
- FreeBSD 12 with OpenSSL 1.1.1b-freebsd  26 Feb 2019 from default installation

### Results
- OSX seems to work fine.
- Ubuntu sometimes works, sometimes it blocks in a way you already observed
- FreeBSD never works: the server always exits with `pthread_create: Too many processes`

**Comment by weinrank** [comment]
I've forgot to mention: I've seen the issue with a single client.
It wasn't necessary to connect with a second client.

**Comment by weinrank** [comment]
We've fixed a bug and added a version check.

The `pthread_create: Too many processes` issue was caused by a mismatch between my OpenSSL version at compile time and the version that was dynamically linked.
I'm quite sure you've trapped into the same problem.
I've added a check which compares the OpenSSL version used at compile time and the version which has been linked.

The "blocking" issue was indeed a bug.
We've set an socket option (SO_REUSEPORT) which behaves differently on FreeBSD and Linux.
I've added an additional check to avoid this unwanted behaviour.

Can you try the fix and report back?
Felix

**Comment by liuqun** [comment]
- OS: Ubuntu 18.04.1 LTS (GNU/Linux 4.15.0-50-generic x86_64)
- OpenSSL version:
```

```

- arguments:
```
liuqun@liuqun-virtual-machine:~/bbb/DTLS-Examples/src$ ./dtls_udp_echo -L 127.0.0.1 -v
Using OpenSSL 1.1.0g  2 Nov 2017
socket: Too many open files
Thread 7f928c440700: done, connection closed.
libgcc_s.so.1 must be installed for pthread_cancel to work
socket: Too many open files
Thread 7f928bc3f700: done, connection closed.
libgcc_s.so.1 must be installed for pthread_cancel to work
Aborted (core dumped)
```

**Comment by weinrank** [comment]
You're right, I were able to reproduce the issue and we'll look into it... thanks!

**Comment by liuqun** [comment]
- Ubuntu 18.04 after upgrade
- OpenSSL lib and header deb package version: `apt show libssl-dev`
```
Package: libssl-dev
Version: 1.1.1-1ubuntu2.1~18.04.1
```

- Output
```
./dtls_udp_echo -L 127.0.0.1 -v
Using OpenSSL 1.1.1  11 Sep 2018
socket: Too many open files
Thread 7fee25c3b700: done, connection closed.
libgcc_s.so.1 must be installed for pthread_cancel to work
Aborted (core dumped)
```

**Comment by weinrank** [comment]
TL;DR: **Use OpenSSL Version 1.1.1a or newer!**

The issue has been reported earlier [DTLSv1_listen() regression on 1.1.x](https://github.com/openssl/openssl/issues/6934) to the OpenSSL project.

Since OpenSSL 1.1.0 only receives security fixes, the issue has only been fixed in 1.1.1a.

We've reproduced the issue with OpenSSL 1.1.0k and 1.1.1 on Linux and FreeBSD.
1.1.1a and higher are working as expected.
